Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,442 @@
+                    m_shapeCache.m_localToWorldDelta, Vector3.zero);
+                // Remember the desired displacement for next frame
+                var prev = m_extra.m_previousDisplacement;
+                m_extra.m_previousDisplacement = displacement;
+
+                if (!VirtualCamera.PreviousStateIsValid || deltaTime < 0 || m_Damping <= 0)
+                    m_extra.m_dampedDisplacement = Vector3.zero;
+                else
+                {
+                    // If a big change from previous frame's desired displacement is detected, 
+                    // assume we are going around a corner and extract that difference for damping
+                    if (Vector2.Angle(prev, displacement) > m_cornerAngleTreshold)
+                        m_extra.m_dampedDisplacement += displacement - prev;
+
+                    m_extra.m_dampedDisplacement -= Damper.Damp(m_extra.m_dampedDisplacement, m_Damping, deltaTime);
+                    displacement -= m_extra.m_dampedDisplacement;
+                }
+                state.PositionCorrection += displacement;
+            }
+        }
+
+        /// <summary>
+        /// Calculates half frustum height for orthographic or perspective camera.
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in CinemachineVirtualCameraBase vcam)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = Mathf.Abs(state.Lens.OrthographicSize);
+            }
+            else
+            {
+                Quaternion inverseRotation = Quaternion.Inverse(m_BoundingShape2D.transform.rotation);
+                Vector3 planePosition = inverseRotation * m_BoundingShape2D.transform.position;
+                Vector3 cameraPosition = inverseRotation * vcam.transform.position;
+                float distance = Mathf.Abs(planePosition.z - cameraPosition.z);
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+            return frustumHeight;
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine, in List<List<Vector2>> pathCache,
+            in Transform localToWorld, in Vector2 offset)
+        {
+            if (ShrinkablePolygon.IsInside(pathCache, positionToConfine, localToWorld, offset))
```

Comment:
This is very inefficient.

Instead of all these matrix multiplies, you should inverse-transform only the camera point - once! - then confine, then transform the result back out - once.

Transforming the camera point includes applying the offset, that way you don't need to do it for every path point.

For efficiency, you should cache both the transform matrix and its inverse, because calculating the matrix inverse is costly.